### ISSUE TITLE
Add more client side validation for job listing submission form

### ISSUE THREAD

**Issue by jom** [Automattic/WP-Job-Manager#1257]
Related to #1163, we should add client side validation to URL/email fields.

**Comment by aheckler** [comment]
cc @tripflex here because this can cause minor data loss when Field Editor is enabled.

For example, if I create a job using just WPJM and enter `example.com` as the application URL, this gets rejected as invalid right away. I am forced to enter a full URL starting with `http://`.

However, if I have Field Editor enabled, the job submission form accepts `example.com`, but then erases it from the "Application Email/URL" field once the job is submitted. 😞 

The customer whose site I can repro this on is running Field Editor 1.4.1, but I have 1.4.6 locally and cannot repro. Did anything change there that could have affected this?

@jom Yeah, we should accept any URL, even bare ones.

**Comment by tripflex** [comment]
@aheckler yeah I kinda mentioned this in the link below and the different options available: https://github.com/Automattic/WP-Job-Manager/pull/1222#issuecomment-341112928

But first thing to point out is that if they're using 1.4.1 ... then that's probably what the problem is .. that version is super duper old ... like almost 2 years old (january 2016), the current release is 1.7.10, and there have been HUGE changes and I doubt a version that old is even compatible with core anymore:
https://plugins.smyl.es/docs/wp-job-manager-field-editor/changelog/

Ive also found that a lot of people who are using that version, as well as 1.6.4, are actually using illegally distributed copies of the plugin, as 

You should also grab the latest version if you're testing as 1.4.6 (may 2016), is also very old:
https://github.com/tripflex/wp-job-manager-field-editor/releases

I don't see anything specific to that version (or between the one they have and one you have) that would cause that issue though.  I'll test on my dev environment and see what happens and report back

**Comment by aheckler** [comment]
@tripflex I don't have access to your GH repos anymore, so can't grab a newer release that way. :/ Not sure if it costs you $$ to add me or not? Maybe the best solution is to create some sort of permanent license + account for us on your site? DM me on our Slack to discuss if you want.

Anyway, I'll tell the user to update and see what happens.

**Comment by tripflex** [comment]
Just tested this myself with 1.7.10 and can't replicate Field Editor allowing `example.com` to be used, the new sanitation handling is still used, but it looks like there's an issue with detecting the URL if a valid URL is not entered.  

https://github.com/Automattic/WP-Job-Manager/blob/master/includes/abstracts/abstract-wp-job-manager-form.php#L310-L317

So it looks like if it doesn't detect it as a URL then it defaults to sanitizing as an email, and in my tests, that just causes an empty value and thus notice on the page saying that field is required (not that it's an invalid format).

Maybe the better way to do this would be to just use `parse_url` and then check the value in `host` key (if it's set), and just check if it's an email, `is_email` before parsing or sanitizing.

_**Either way,**_ if the user does not enter a valid URL, when the link is output it's going to link to a subdirectory of the user's site.  So if `example.com` was allowed to be used and saved, when that link is output as

```html
<a href="example.com">contact</a>
```

That's actually going to link to `http://listingsite.com/example.com`

Here's the results from `parse_url` when not specifying `PHP_URL_HOST`:

```
Value: //google.com Result:array(1) {
  ["host"]=>
  string(10) "google.com"
}
Value: http://google.com  Result:array(2) {
  ["scheme"]=>
  string(4) "http"
  ["host"]=>
  string(10) "google.com"
}
Value: https://google.com  Result:array(2) {
  ["scheme"]=>
  string(5) "https"
  ["host"]=>
  string(10) "google.com"
}
Value: google.com  Result:array(1) {
  ["path"]=>
  string(10) "google.com"
}
Value: some@email.com  Result:array(1) {
  ["path"]=>
  string(14) "some@email.com"
}
```
Here's a glot if you want to play around with `parse_url`:
https://glot.io/snippets/ew65bykyok

**Comment by wpfan2099** [comment]
hello dear all - i really like this idea!!! Many thanks for coming up with this great idea and approach.

**Comment by codestor4** [comment]
1228036-zen

User requests:

1. validate all client side (javascript) 
2. show all issues at once (instead of one by one where user fills one field and tries to submit and then it tells us some other field is still empty and so on).
